### ISSUE TITLE
dials.find_spots fails in 3.14.1

### ISSUE THREAD

**Issue by KevinBattaile** [dials/dials#2401]
I'm running on CentOS stream 8 and when I try to index a group of 3 consecutive images off an Eiger2 (h5 format) I get the error pasted below. This seems to work fine in 3.13.0 but started with 3.14.0.

Command list:

1. dials.import allow_multiple_sequences=true 1/PSL-10_9/PSL-1009_626_master.h5 2/PSL-10_9/PSL-1009_625_master.h5
2. dials.find_spots imported.expt

-----------------------------
```
[kbattaile@Nostromo PSL-1009]$ dials.find_spots imported.expt
DIALS (2018) Acta Cryst. D74, 85-97. https://doi.org/10.1107/S2059798317017235
DIALS 3.14.1-ge7f638df6-release
The following parameters have been modified:

input {
  experiments = imported.expt
}

Setting spotfinder.filter.min_spot_size=3
Configuring spot finder from input parameters
--------------------------------------------------------------------------------
Finding strong spots in imageset 0
--------------------------------------------------------------------------------

Finding spots in image 1 to 3...
Setting nproc=28
Setting chunksize=1
Extracting strong pixels from images
 Using multiprocessing with 3 parallel job(s)



easy_mp crash detected; subprocess trace: ----
Traceback (most recent call last):
  File "/data/Software/DIALS/dials-v3-14-1/modules/cctbx_project/libtbx/scheduling/job_scheduler.py", line 74, in job_cycle
    value = target( *args, **kwargs )
  File "/data/Software/DIALS/dials-v3-14-1/modules/dials/src/dials/util/mp.py", line 114, in __call__
    return [self.__function(item) for item in iterable]
  File "/data/Software/DIALS/dials-v3-14-1/modules/dials/src/dials/util/mp.py", line 114, in <listcomp>
    return [self.__function(item) for item in iterable]
  File "/data/Software/DIALS/dials-v3-14-1/modules/dials/src/dials/algorithms/spot_finding/finder.py", line 260, in __call__
    result = self.function(task)
  File "/data/Software/DIALS/dials-v3-14-1/modules/dials/src/dials/algorithms/spot_finding/finder.py", line 78, in __call__
    image = self.imageset.get_corrected_data(index)
  File "/data/Software/DIALS/dials-v3-14-1/modules/dxtbx/src/dxtbx/format/FormatMultiImage.py", line 47, in read
    return format_instance.get_raw_data(index)
  File "/data/Software/DIALS/dials-v3-14-1/modules/dxtbx/src/dxtbx/format/FormatNXmxEigerFilewriter.py", line 50, in get_raw_data
    raw_data = get_raw_data(nxdata, nxdetector, index)
  File "/data/Software/DIALS/dials-v3-14-1/modules/dxtbx/src/dxtbx/format/FormatNXmxEigerFilewriter.py", line 88, in get_raw_data
    data_as_flex = _dataset_as_flex(sliced_outer, tuple(module_slices))
  File "/data/Software/DIALS/dials-v3-14-1/modules/dxtbx/src/dxtbx/nexus/__init__.py", line 523, in _dataset_as_flex
    raise TypeError(f"Unsupported integer dtype {data_np.dtype}")
Unsupported integer dtype uint32
----------------------------------------------


Traceback (most recent call last):
  File "/data/Software/DIALS/dials-v3-14-1/build/../modules/dials/src/dials/command_line/find_spots.py", line 242, in <module>
    run()
  File "/data/Software/DIALS/dials-v3-14-1/conda_base/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/data/Software/DIALS/dials-v3-14-1/build/../modules/dials/src/dials/command_line/find_spots.py", line 236, in run
    results = do_spotfinding(experiments, params)
  File "/data/Software/DIALS/dials-v3-14-1/build/../modules/dials/src/dials/command_line/find_spots.py", line 127, in do_spotfinding
    reflections = flex.reflection_table.from_observations(experiments, params)
  File "/data/Software/DIALS/dials-v3-14-1/modules/dials/src/dials/array_family/flex_ext.py", line 183, in from_observations
    return spotfinder.find_spots(experiments)
  File "/data/Software/DIALS/dials-v3-14-1/modules/dials/src/dials/algorithms/spot_finding/finder.py", line 704, in find_spots
    table, hot_mask = self._find_spots_in_imageset(imageset)
  File "/data/Software/DIALS/dials-v3-14-1/modules/dials/src/dials/algorithms/spot_finding/finder.py", line 817, in _find_spots_in_imageset
    r, h = extract_spots(imageset[j0:j1])
  File "/data/Software/DIALS/dials-v3-14-1/modules/dials/src/dials/algorithms/spot_finding/finder.py", line 417, in __call__
    return self._find_spots(imageset)
  File "/data/Software/DIALS/dials-v3-14-1/modules/dials/src/dials/algorithms/spot_finding/finder.py", line 508, in _find_spots
    batch_multi_node_parallel_map(
  File "/data/Software/DIALS/dials-v3-14-1/modules/dials/src/dials/util/mp.py", line 194, in batch_multi_node_parallel_map
    return multi_node_parallel_map(
  File "/data/Software/DIALS/dials-v3-14-1/modules/dials/src/dials/util/mp.py", line 165, in multi_node_parallel_map
    result = libtbx.easy_mp.parallel_map(
  File "/data/Software/DIALS/dials-v3-14-1/modules/cctbx_project/libtbx/easy_mp.py", line 637, in parallel_map
    result = res()
  File "/data/Software/DIALS/dials-v3-14-1/modules/cctbx_project/libtbx/scheduling/result.py", line 119, in __call__
    self.traceback( exception = self.exception() )
  File "/data/Software/DIALS/dials-v3-14-1/modules/cctbx_project/libtbx/scheduling/stacktrace.py", line 117, in __call__
    self.raise_handler( exception = exception )
  File "/data/Software/DIALS/dials-v3-14-1/modules/cctbx_project/libtbx/scheduling/stacktrace.py", line 134, in raise_with_traceback
    raise_(type(exception), exception, self.traceback)
  File "/data/Software/DIALS/dials-v3-14-1/conda_base/lib/python3.10/site-packages/six.py", line 719, in reraise
    raise value
  File "/data/Software/DIALS/dials-v3-14-1/modules/cctbx_project/libtbx/scheduling/mainthread.py", line 100, in poll
    value = target( *args, **kwargs )
  File "/data/Software/DIALS/dials-v3-14-1/modules/dials/src/dials/util/mp.py", line 88, in __call__
    return libtbx.easy_mp.parallel_map(
  File "/data/Software/DIALS/dials-v3-14-1/modules/cctbx_project/libtbx/easy_mp.py", line 637, in parallel_map
    result = res()
  File "/data/Software/DIALS/dials-v3-14-1/modules/cctbx_project/libtbx/scheduling/result.py", line 119, in __call__
    self.traceback( exception = self.exception() )
  File "/data/Software/DIALS/dials-v3-14-1/modules/cctbx_project/libtbx/scheduling/stacktrace.py", line 117, in __call__
    self.raise_handler( exception = exception )
  File "/data/Software/DIALS/dials-v3-14-1/modules/cctbx_project/libtbx/scheduling/stacktrace.py", line 147, in raise_with_stacktrace
    self.stacktrace( exception = exception )
  File "/data/Software/DIALS/dials-v3-14-1/modules/cctbx_project/libtbx/scheduling/stacktrace.py", line 88, in __call__
    raise exception
TypeError: Please report this error to dials-support@lists.sourceforge.net: Unsupported integer dtype uint32
```

**Comment by ndevenish** [comment]
Hi Kevin,

This is happening because the internal data type is 32-bit unsigned, but DIALS is trying to use 32-bit signed. It's assuming that it is unsafe to convert 32bit unsigned→32-bit signed unless the [bit_depth_readout](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-bit-depth-readout-field) NXmx field is set to 32.

On the other hand, the NXmx spec doesn't appear to have a way to signify signed or unsigned, so maybe we need to consistently assume that one of these conversions are safe. @graeme-winter, do you have any thoughts on the best way to proceed here?

**Comment by graeme-winter** [comment]
Thanks for the reminder and call in here.

On a fundamental (i.e. signal processing level) the _counts_ are all unsigned ints - either `uint32_t` or `uint16_t` for faster frame rates (ignoring the `uint8_t` option for the moment)

However, from a symbolic perspective the detector data also use `0xffffffff` and `0xfffffffe` as "masked" pixels which are - if interpreted as `int32_t` - negative numbers as -1 and -2 respectively. However - the data should not be interpreted as -ve, and for counting data we have a trusted range from `[0...max)` I think, and the symbols are outside of this range however they are interpreted. 

We cannot have negative values. 

So, quick solution here is to ignore the fact that strictly interpreting the symbols as `int32` or `uint32` is wrong and simply accept that either numerical interpretation gives the same effective answer, as in the bits in the file to the meaning of those bits. TL;DR - interpreting `uint32_t` data as `int32_t` will be fine for an Eiger for all circumstances I am aware of.

**Comment by graeme-winter** [comment]
@kmdalton any way we could get the data files? That would probably help us come up with a good solution to the problem...

**Comment by graeme-winter** [comment]
@KevinBattaile any chance we could get the data files? Would be very useful for working out how best to fix this.

**Comment by KevinBattaile** [comment]
I could certainly get you some data to work with, just let me know the best way to do so.

Any idea why this became a problem with the 3.14.x version?

**Comment by graeme-winter** [comment]
Probably became an issue with 3.14 versions as we overhauled the handling of NeXus data in the relatively recent history (I can't find the PR right now, but I remember it happening towards the end of last year)

Data -> any chance you can get a google drive? Or I can probably email you a link to a writable onedrive folder

**Comment by KevinBattaile** [comment]
@graeme-winter what is an e-mail address I can use to send  you a one drive link for a few data frames?

**Comment by graeme-winter** [comment]
> @graeme-winter what is an e-mail address I can use to send you a one drive link for a few data frames?

graeme.winter [@] diamond.ac.uk

**Comment by graeme-winter** [comment]
OK, I got this. Fix may not be instant.

**Comment by graeme-winter** [comment]
@KevinBattaile  OK thank you for sending the data

With the two change sets linked above the spot finding now works sensibly:

```DIALS (2018) Acta Cryst. D74, 85-97. https://doi.org/10.1107/S2059798317017235
DIALS 3.dev.968-gf11792932
The following parameters have been modified:

spotfinder {
  mp {
    nproc = 1
  }
}
input {
  experiments = imported.expt
}

Setting spotfinder.filter.min_spot_size=3
Configuring spot finder from input parameters
--------------------------------------------------------------------------------
Finding strong spots in imageset 0
--------------------------------------------------------------------------------

Finding spots in image 1 to 3...
Setting chunksize=3
Extracting strong pixels from images
 Using multiprocessing with 1 parallel job(s)

Found 7554 strong pixels on image 1
Found 9153 strong pixels on image 2
Found 8683 strong pixels on image 3

Extracted 707 spots
Removed 170 spots with size < 3 pixels
Removed 0 spots with size > 1000 pixels
Calculated 537 spot centroids
Calculated 537 spot intensities
Filtered 410 of 537 spots by peak-centroid distance

Histogram of per-image spot count for imageset 0:
410 spots found on 3 images (max 162 / bin)
  *
  *
 **
***
***
***
***
***
***
***
1 3

--------------------------------------------------------------------------------
Saved 410 reflections to strong.refl
```

Now to get the changes merged, though I think the NXmx one could be a pain as something went wrong in the robotic nonsense...

**Comment by graeme-winter** [comment]
Fix in cctbx/dxtbx#632 should be a quick and easy one, you could probably drop this over the top of the same file in your install @KevinBattaile and everything should "just work"

YMMV of course

**Comment by ndevenish** [comment]
Hi @KevinBattaile, this should be fixed in DIALS 3.14.2, which we just released. Please let me know if you have any issues with it.
